File: java/Ice/async/HelloI.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

import Demo.*;

public class HelloI implements Hello
{
    public HelloI(WorkQueue workQueue)
    {
        _workQueue = workQueue;
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> sayHelloAsync(int delay, com.zeroc.Ice.Current current)
    {
        if(delay == 0)
        {
            System.out.println("Hello World!");
            return java.util.concurrent.CompletableFuture.completedFuture((Void)null);
        }
        else
        {
            java.util.concurrent.CompletableFuture<Void> r = new java.util.concurrent.CompletableFuture<>();
            _workQueue.add(r, delay);
            return r;
        }
    }

    @Override
    public void shutdown(com.zeroc.Ice.Current current)
    {
        System.out.println("Shutting down...");

        _workQueue._destroy();
        current.adapter.getCommunicator().shutdown();
    }

    private WorkQueue _workQueue;
}

File: cpp/Ice/async/HelloI.cpp
// Copyright (c) ZeroC, Inc.

#include "HelloI.h"
#include <Ice/Ice.h>
#include <iostream>

using namespace std;

HelloI::HelloI(const shared_ptr<WorkQueue>& workQueue) : _workQueue(workQueue) {}

void
HelloI::sayHelloAsync(int delay, function<void()> response, function<void(exception_ptr)> error, const Ice::Current&)
{
    if (delay == 0)
    {
        cout << "Hello World!" << endl;
        response();
    }
    else
    {
        _workQueue->add(delay, std::move(response), std::move(error));
    }
}

void
HelloI::shutdown(const Ice::Current& curr)
{
    cout << "Shutting down..." << endl;

    _workQueue->destroy();
    curr.adapter->getCommunicator()->shutdown();
}
